shell/platform/common/accessibility_bridge.cc

```diff
@@ -177,6 +199,59 @@ void AccessibilityBridge::OnAtomicUpdateFinished(
   }
 }
 
+std::optional<ui::AXTreeUpdate>
+AccessibilityBridge::CreateRemoveReparentedNodesUpdate() {
+  std::unordered_map<int32_t, ui::AXNodeData> updates;
+
+  for (auto node_update : pending_semantics_node_updates_) {
+    for (int32_t child_id : node_update.second.children_in_traversal_order) {
+      // Skip nodes that don't exist or have a parent in the current tree.
+      ui::AXNode* child = tree_.GetFromId(child_id);
+      if (!child || !child->parent()) {
+        continue;
+      }
+
+      // Skip nodes whose parents are unchanged.
+      if (child->parent()->id() == node_update.second.id) {
+        continue;
+      }
+
+      // This pending update moves the current child node.
+      // That new child must have a corresponding pending update.
+      assert(pending_semantics_node_updates_.find(child_id) !=
```

Comment:
Does it need to have a update? a child can be moved to a different parent without update its content i think?

Reply:
It also probably worth check that its old parent must have an update in pending_semantics_node_updates_

Reply:
We've hit this in a previous (internal) chrome-based embedder; when a child is removed from a node, we need to send the entire subtree of its parent.

Reply:
> Does it need to have a update? a child can be moved to a different parent without update its content i think?

The only way to move a node is by deleting it first. Thus, a node and its full subtree must be in the update if they are moved. If you don't include the moved node, the `AXTree` will reject the update as some parent node references a child that wasn't included in the update.

> It also probably worth check that its old parent must have an update in pending_semantics_node_updates_

Good idea, will add!

Reply:
maybe have a global key text widget moving around in different frame? In that case I think you will only receive parent update.

and I just realized the parent may not have an update if the parent itself is remove from the tree entirely

Reply:
I verified that moving widgets in this app sends updates for the moved widget's entire subtree (instead of just the moved widget):

<details>
<summary>Swap nodes with global keys</summary>

```dart
import 'package:flutter/material.dart';

void main() => runApp(const MyApp());

class MyApp extends StatefulWidget {
  const MyApp({Key? key}) : super(key: key);

  @override
  State<MyApp> createState() => _MyAppState();
}

class _MyAppState extends State<MyApp> {
  final GlobalKey _middle1Key = GlobalKey();
  final GlobalKey _middle2Key = GlobalKey();
  final GlobalKey _leaf1Key = GlobalKey();
  final GlobalKey _leaf2Key = GlobalKey();

  bool _swap = false;

  @override
  Widget build(BuildContext context) {
    final middle1 = Semantics(
      key: _middle1Key,
      container: true,
      explicitChildNodes: true,
      child: Text('Leaf 1 text', key: _leaf1Key),
    );

    final middle2 = Semantics(
      key: _middle2Key,
      container: true,
      explicitChildNodes: true,
      child: Text('Leaf 2 text', key: _leaf2Key),
    );

    return WidgetsApp(
      color: const Color(0xFF000000),
      home: Row(children: [
        Semantics(
          explicitChildNodes: true,
          label: 'Top 1 label',
          child: !_swap ? middle1 : middle2,
        ),
        Semantics(
          explicitChildNodes: true,
          label: 'Top 2 label',
          child: !_swap ? middle2 : middle1,
        ),
        TextButton(
          child: const Text('Swap'),
          onPressed: () {
            setState(() {
              _swap = !_swap;
            });
          },
        )
      ]),
      pageRouteBuilder: <T>(RouteSettings settings, WidgetBuilder builder) => PageRouteBuilder<T>(
        settings: settings,
        pageBuilder: (BuildContext context, Animation<double> animation, Animation<double> secondaryAnimation) => builder(context),
      ),
    );
  }
}
```

</details>

Reply:
Marking as resolved, but let me know if you have additional feedback!